pyro/distributions/zero_inflated.py

```diff
@@ -70,17 +71,29 @@ def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
 
+        support = self.support
+        epsilon = abs(torch.finfo(value.dtype).eps)
+        zero_idx = (value == 0)
+
+        if hasattr(support, "lower_bound"):
+            if support.lower_bound == 0.0:
+                value = value.clamp_min(epsilon)
+
+        if hasattr(support, "upper_bound"):
+            if support.upper_bound == 1.0:
+                value = value.clamp_max(1 - epsilon)
```

Comment:
Let's see, is this sufficiently general?  🤔 

I believe `.upper_bound` and `.lower_bound` can be tensors, which would break this logic.  I guess one workaround would be to use our `is_identically_*()` helpers
```python
from pyro.distributions.util import is_identically_one, is_identically_zero

if is_identically_zero(getattr(support, "lower_bound", None)):  ...
if is_identically_one(getattr(support, "upper_bound", None)):  ...
```
I'm not sure how ZeroInflatedDistribution is used, but I guess you could even allow other scalar bounds via
```python
lb = getattr(support, "lower_bound", None)
if isinstance(lb, (float, int)):
    value = value.clamp_min(lb + epsilon)
ub = ...
```

Another thought is that these zero-inflated distributions should only add a new point at zero (e.g. changing support from `(0,1)` to `[0,1)` or `(-1,0)` to `(-1,0]`), so maybe we should use `is_identically_zero` for both the lower and upper bound checks?

WDYT?

Reply:
> from pyro.distributions.util import is_identically_one, is_identically_zero
> if is_identically_zero(getattr(support, "lower_bound", None)):  ...
> if is_identically_one(getattr(support, "upper_bound", None)):  ...

These are actuall good suggestions. But the problem is that not all distributions have a lower/upper bound in their support, e.g., Normal. So we first need to check if it has such a support. This is true for, e.g., the `Beta`.

The same goes for your second comment.